File: csniffer/packetsniffer.c
#include "headers.h"

typedef struct{
    unsigned char dest_mac[6];
    unsigned char src_mac[6];
    uint16_t proto;
}headers; 

typedef struct{
    uint8_t version_ihl;
    uint8_t tos;
    uint16_t total_len;
    uint8_t ip_proto;
    unsigned char src_ip[4];
    unsigned char dst_ip[4];
}ip_data;

typedef struct{
    uint16_t src_port;
    uint16_t dst_port;
}udppayload;

typedef struct{
    uint8_t icmp_type;
    uint8_t icmp_code;
    uint16_t icmp_checksum;
}icmppayload;

typedef struct{
    uint16_t src_port;
    uint16_t dst_port;
    uint16_t offset_flags;
}tcppayload;

int main(){
    struct sockaddr *addr;

    headers head;

    struct sockaddr_storage localaddr;
    
    socklen_t addrsize = sizeof(localaddr);

    int sock = socket(AF_PACKET, SOCK_RAW, htons(0x0003));

    if(sock == -1){
        perror("sock");
    }
    
    unsigned char buffer[65535];

    while(1){
        
        ssize_t bytes_received = 0;

        ip_data ipdata;

        bytes_received = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr *)&localaddr, &addrsize);
        
        memcpy(&head.dest_mac, buffer, sizeof(head.dest_mac));
        memcpy(&head.src_mac, buffer + 6, sizeof(head.src_mac));

        memcpy(&head.proto, buffer + 12, sizeof(head.proto));

        head.proto = ntohs(head.proto);

        //printf("Source: %x Destination: %x Proto: %x\n", head.src_mac, head.dest_mac, head.proto);

        if(head.proto == 0x0800){

            uint8_t version;
            uint8_t ihl;
            int ip_header_len;
            char src_ip_readable[INET_ADDRSTRLEN];
            char dst_ip_readable[INET_ADDRSTRLEN];
            
            memcpy(&ipdata.version_ihl, buffer + 14, sizeof(ipdata.version_ihl));
            memcpy(&ipdata.src_ip, buffer + 26, sizeof(ipdata.src_ip));
            memcpy(&ipdata.dst_ip, buffer + 30, sizeof(ipdata.dst_ip));
            memcpy(&ipdata.total_len, buffer + 16, sizeof(ipdata.total_len));
            memcpy(&ipdata.ip_proto, buffer + 23, sizeof(ipdata.ip_proto));

            ipdata.total_len = ntohs(ipdata.total_len);

            version = ipdata.version_ihl >>4;
            ihl = ipdata.version_ihl & 0x0F;
            ip_header_len = ihl * 4;

            inet_ntop(AF_INET, ipdata.src_ip, src_ip_readable, sizeof(src_ip_readable));
            inet_ntop(AF_INET, ipdata.dst_ip, dst_ip_readable, sizeof(dst_ip_readable));

            int offset = 14 + ip_header_len;
            
            if (ipdata.total_len < offset){
                continue;
            }

            size_t transport_length = ipdata.total_len - offset;

            unsigned char *transport_data = malloc(transport_length);

            if(transport_data == NULL){
                continue;
            }

            //printf("Transport Length: %d\n", transport_length);
            //printf("IP Proto: %d\n", ipdata.ip_proto);

            memcpy(transport_data, buffer + offset, transport_length);

            //printf("Transport Data: %p\n", (void*)transport_data);

            if(ipdata.ip_proto == 17 && transport_length >= 8){
                udppayload payload;

                memcpy(&payload.src_port, transport_data, sizeof(payload.src_port));
                memcpy(&payload.dst_port, transport_data + 2, sizeof(payload.dst_port));

                payload.src_port = ntohs(payload.src_port);
                payload.dst_port = ntohs(payload.dst_port);

               printf("Source IP: %s -> Destination IP: %s Source Port: %d - > Destination Port: %d\n", src_ip_readable, dst_ip_readable, payload.src_port, payload.dst_port);
            
            }

            if(ipdata.ip_proto == 1 && transport_length >= 4){
                icmppayload payload;

                memcpy(&payload.icmp_type, transport_data, sizeof(payload.icmp_type));
                memcpy(&payload.icmp_code, transport_data + 1 , sizeof(payload.icmp_code));

                payload.icmp_type = ntohs(payload.icmp_type);
                payload.icmp_code = ntohs(payload.icmp_code);

                printf("Source IP: %s -> Destination IP: %s ICMP type: %d ICMP code: ", src_ip_readable, dst_ip_readable, payload.icmp_type, payload.icmp_code);
                
            }


            if(ipdata.ip_proto == 6 && transport_length >= 20){
                tcppayload payload;
                uint16_t tcp_offset, urg_flag, ack_flag, psh_flag, rst_flag, syn_flag, fin_flag;
                int count = 0;
                char *flags[64];
                const char* srcservice;
                const char* dstservice;

                memcpy(&payload.src_port, transport_data, sizeof(payload.src_port));
                memcpy(&payload.dst_port, transport_data + 2, sizeof(payload.dst_port));
                memcpy(&payload.offset_flags, transport_data + 10, sizeof(payload.offset_flags));

                payload.src_port = ntohs(payload.src_port);
                payload.dst_port = ntohs(payload.dst_port);

                payload.offset_flags = ntohs(payload.offset_flags);
                tcp_offset = (payload.offset_flags >> 12) * 4;
                
                urg_flag = (payload.offset_flags & 32);
                ack_flag = (payload.offset_flags & 16);
                psh_flag = (payload.offset_flags & 8); 
                rst_flag = (payload.offset_flags & 4); 
                syn_flag = (payload.offset_flags & 2); 
                fin_flag = (payload.offset_flags & 1);

                if(urg_flag != 0){
                    flags[count] = "URG";
                    count++;
                }
                if(ack_flag != 0){
                    flags[count] = "ACK";
                    count++;
                }

                if(psh_flag != 0){
                    flags[count] = "PSH";
                    count++;
                }

                if(rst_flag != 0){
                    flags[count] = "RST";
                    count++;
                }

                if(syn_flag != 0){
                    flags[count] = "SYN";
                    count++;
                }

                if(fin_flag != 0){
                    flags[count] = "FIN";
                    count++;
                }
        
                srcservice = TCP_Service(payload.src_port);
                dstservice = TCP_Service(payload.dst_port);
                
                printf("%s -> %s %d -> %d %s %s", src_ip_readable, dst_ip_readable, payload.src_port, payload.dst_port, srcservice, dstservice);
                for(int i=0; i < count; i++){
                    printf("%s ", flags[i]);
                }
                printf("\n");


            }

            free(transport_data);
             
        }
}
}

File: pythonsniffer/packetsniffer.py
from utilities import *

sock = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(0x0003))

active_ifaces = get_active_interface()

if not active_ifaces:
    print("No active network interface")
    exit(1)
target_iface = active_ifaces[0].strip('\x00')

sock.bind((target_iface, 0))

print(f"PACKET SNIFFER RUNNING AT {target_iface}...")

while True:
    raw_bytes, addr = sock.recvfrom(65535)
    headers = raw_bytes[:14]
    dest_mac, src_mac, proto = struct.unpack('!6s6sH', headers) #6s for 6 bytes

    if proto == 0x0800: #Only accepting IPv4

        ip_data = raw_bytes[14:]
        version_ihl, tos, total_len, idfication, flags_fragment, ttl, ip_proto,check_sum, src_ip, dst_ip = struct.unpack("!BBHHHBBH4s4s", ip_data[:20])

        version = version_ihl >> 4  
        ihl = version_ihl & 0x0F 
        ip_header_len = ihl * 4

        transport_data = ip_data[ip_header_len:]

        src_ip_readeable = socket.inet_ntoa(src_ip)
        dst_ip_readeable = socket.inet_ntoa(dst_ip)

        if ip_proto == 17 and len(transport_data) >= 8: #Getting UDP

            src_port, dst_port, udp_len, udp_checksum = struct.unpack("!HHHH", transport_data[:8]) #First 8 bytes is the header
            
            udp_payload = transport_data[8:] 

            service = UDP_SERVICES.get(src_port) or UDP_SERVICES.get(dst_port) or "UNKNOWN"

            if service == "DNS" and len(udp_payload) >= 12:
                dns_id, dns_flags, qdcount, anscount, nscount, arcount = struct.unpack("!HHHHHH", udp_payload[:12])
                
                is_response = dns_flags & 0x8000 

                offset = 12 # To get to first byte
                labels = []
                
                while True: # Getting the domain from the payload
                    dns_length = udp_payload[offset]

                    if dns_length == 0:
                        break

                    label = udp_payload[offset + 1 : offset + dns_length + 1].decode() 
                    labels.append(label)

                    offset += 1 + dns_length
                
                offset += 1
                qtype, qclass = struct.unpack("!HH", udp_payload[offset:offset+4]) #Query payload

                domain = ".".join(labels)

                qtype_text = DNS_TYPES.get(qtype, str(qtype))

                offset += 4
                
                if qclass == 1:
                    qclass_text = "IN"
                else:
                    qclass_text = str(qclass)

                if anscount > 0:
                    if udp_payload[offset] & 0xC0 == 0xC0:

                        offset += 2
                        atype, aclass, ttl,rdlength = struct.unpack("!HHIH", udp_payload[offset:offset+10])
                        offset += 10
                        print("RDLENGTH:",rdlength)
                        rdata = udp_payload[offset: offset+rdlength]

                        if atype == 1 and rdlength == 4:
                            ipv4addr = ".".join(map(str, rdata))
                            print("ANSWER IPV4", "TYPE:", atype, "CLASS:", aclass, "TTL:", ttl, "IP:", ipv4addr)

                        elif atype == 28 and rdlength == 16:
                            groups = []
                            for i in range(0,16,2):
                                group = rdata[i:i+2]
                                groups.append(group.hex())
                    
                            ipv6addr = ipaddress.IPv6Address(":".join(groups))
                            
                            print("ANSWER IPV6", "TYPE:", atype, "CLASS:", aclass, "TTL:", ttl, "IP:", ipv6addr)
                        else:
                            print("ANSWER", "TYPE:", atype, "CLASS:", aclass, "TTL:", ttl, "RDATA", rdata)


        
                if is_response:
                    print("RESPONSE","ID:", dns_id, domain, qtype_text, qclass_text)
                else:
                    print("QUERY", "ID:", dns_id, domain, qtype_text, qclass_text)

                #print(src_ip_readeable, "->", dst_ip_readeable, src_port, "->", dst_port, service)


#        if ip_proto == 1 and len(transport_data) >= 4:

#            icmp_type, icmp_code, icmp_checksum = struct.unpack("!BBH", transport_data[:4])

#            print(src_ip_readeable, dst_ip_readeable, format_icmp(icmp_type), icmp_code)

        if ip_proto == 6 and len(transport_data) >= 20:
            src_port, dst_port, seq, ack, offset_reserved_flags, window, tcp_checksum, urg_ptr = struct.unpack('!HHLLHHHH', transport_data[:20])
            
            flags = []

            tcp_offset = (offset_reserved_flags >> 12) * 4

            
            tcp_payload = transport_data[tcp_offset:]

            urg_flag = (offset_reserved_flags & 32) >> 5
            ack_flag = (offset_reserved_flags & 16) >> 4
            psh_flag = (offset_reserved_flags & 8) >> 3
            rst_flag = (offset_reserved_flags & 4) >> 2
            syn_flag = (offset_reserved_flags & 2) >> 1
            fin_flag = offset_reserved_flags & 1

            if syn_flag:
                flags.append("SYN")
            if ack_flag:
                flags.append("ACK")
            if fin_flag:
                flags.append("FIN")
            if rst_flag:
                flags.append("RST")
            if psh_flag:
                flags.append("PSH")
            if urg_flag:
                flags.append("URG")

            service = TCP_SERVICES.get(src_port) or TCP_SERVICES.get(dst_port) or "UNKNOWN"
            preview = tcp_payload[:30]
            first3 = tcp_payload[:3]

            if len(tcp_payload) > 0:

                if first3[:2] in (b'\x16\x03', b'\x17\x03', b'\x14\x03'):
                    print("Likely TLS traffic")

                elif any(preview.startswith(prefix) for prefix in HTTP_PREFIXES):

                    print("HTTP", repr(preview.decode(errors='ignore')))

                elif preview.startswith(b'SSH-'):

                    print("SSH", repr(preview.decode(errors='ignore')))

                else: 
                    payload_text = tcp_payload[:64].decode(errors='ignore')

                    if is_readable(payload_text):

                        print("TEXT", repr(payload_text))
                        
                    else:

                        print("HEX", preview.hex())

            #print(src_ip_readeable, "->", dst_ip_readeable, src_port, "->", dst_port, flags, service, "Payload len: ", len(tcp_payload))
